enhancements/machine-api/machine-deletion-hooks.md

```diff
@@ -0,0 +1,412 @@
+creation, sometime after Machine creation by a user, or sometime after Machine
+creation by another controller or application.
+
+## Motivation
+
+Allow custom and 3rd party components to easily interact with a Machine or
+related resources while that Machine's reconciliation is temporarily paused.
+This pause in reconciliation will allow these custom components to take action
+after a Machine has been marked for deletion, but prior to the Machine being
+drained and/or associated instance terminated.
+
+In particular, this will allow the Etcd Operator to prevent a Control Plane
+Machine from being drained/removed until the Etcd Operator has had a chance
+to synchronise the Etcd data to the replacement Machine.
+
+This could also be used by other components such as Storage to allow them to
+ensure safe detachment of volumes before the Machine is deleted.
+
+### Goals
+
+- Define an initial set of hook points for the deletion phase.
+- Define an initial set and form of related annotations.
+- Define basic expectations for a controller or process that responds to a
+lifecycle hook.
+
+### Non-Goals
+
+- Create an exhaustive list of hooks; we can add more over time.
+- Create new machine phases.
+- Create a mechanism to signal what lifecycle point a machine is at currently.
+- Dictate implementation of controllers that respond to the hooks.
+- Implement ordering in the machine controller.
+- Require anyone to use these hooks for normal machine operations, these are
+strictly optional and for custom integrations only.
+
+## Proposal
+
+- Utilize annotations to implement lifecycle hooks.
+- Each lifecycle point can have 0 or more hooks.
+- Hooks do not enforce ordering.
+- Hooks found during machine reconciliation effectively pause reconciliation
+until all hooks for that lifecycle point are removed from a machine's annotations.
+
+### User Stories
+
+#### Story 1
+(pre-terminate) As an operator, I would like to have the ability to perform
+different actions between the time a Machine is marked deleted in the API and
+the time the Machine is deleted from the infrastructure provider.
+
+For example, when replacing a Control Plane Machine, ensure a new Control
+Plane Machine has been successfully created and joined to the cluster before
+removing the instance of the deleted Machine. This might be useful in case
+there are disruptions during replacement and we need the disk of the existing
+instance to perform some disaster recovery operation.  This will also prevent
+prolonged periods of having one fewer Control Plane host in the event the
+replacement instance does not come up in a timely manner.
+
+#### Story 2
+(pre-terminate) As an operator, I would to have the ability to perform some
+action between the time that a Machine has been drained, but before it is
+removed from the infrastructure provider.
+
+For example, when storage is attached to the Machine due to some pod placement,
```

Comment:
This is generic enough as it should be built in. https://github.com/kubernetes-sigs/cluster-api/pull/4945

Reply:
Yes this is possibly not the best example, though was just copied across from the upstream proposal. I wonder if a better example might be something to do with GPUs and making sure they are detached properly, as this AFAIK is not a first party API field that can be observed in the same way that volume attachments can be.

CC @elmiko do you think GPUs might be a good example here?

Reply:
interesting question, i haven't come across cases where _detaching_ the gpu has been an issue. most of the problems we see are around synchronizing the attachment/driver deployment for gpus. this might be worth looking into on the platforms that use post-creation gpu attachment.

Reply:
One use case idea I've seen in the past, when using an ELK/EFK stack, the logging daemon on the Node can become overwhelmed if particular applications are producing excessive logs. To ensure that none are lost when the node is removed, a logging operator may want to add a delay post drain to allow the logging daemon on the Node to catch up (ie once new log lines aren't being produced, it should be able to catch up). I'll add this as an extra example

Reply:
that sounds like a solid example to me